### PR TITLE
libpng 1.5.27

### LINKED ISSUE #6694: OpenCV 2.x uses vulnerable version of libpng
OpenCV 2.x (or at least the Android SDK version) uses a vulnerable version of libpng. Any applications uploaded to Google Play are given a warning and directed to this page: https://support.google.com/faqs/answer/7011127

The vulnerability in question is https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-8540, which stems from an out-of-bounds memory access that could potentially lead to code execution Versions 1.0.x before 1.0.66, 1.1.x and 1.2.x before 1.2.56, 1.3.x and 1.4.x before 1.4.19 and 1.5.x before 1.5.26 are affected.

Google is going to block apps using vulnerable versions of libpng after 17 Sep 2016.

### PR DESCRIPTION
**Pull request by alalek**
resolves #6694

### CODE REVIEW

File: 3rdparty/readme.txt [hunk 1]
```diff
@@ -15,7 +15,7 @@ libjpeg 8d (8.4)  -   The Independent JPEG Group's JPEG software.
                       On UNIX systems configure script takes care of it.
 ------------------------------------------------------------------------------------
 libpng 1.5.12     -   Portable Network Graphics library.
```
Comment (apavlenko) [inline 3rdparty/readme.txt]: shouldn't `libpng 1.5.12` be changed to `libpng 1.5.27`?

**Comment by apavlenko** [comment]
:+1:

**Comment by boazgarty** [comment]
Hi,
can you please add the fix based on the official openCV 2.4.11 release.
I have many compilation errors while trying to build the night build from the 1 to July and I have problems publish my app to google play due to "libpng containing a security vulnerability".

Thanks!

**Comment by nisrulz** [comment]
Can someone here help me in outlining how to build android sdk for `2.4` branch from the latest source code, so as to be able to get the `libpng 1.5.27` patch.

I need to update my application on Playstore and I donot wish to go to the route of having to use the Android Manager App. 

Any help or direction would be helpful.

**Comment by alalek** [comment]
@nisrulz There is patch #7289 which adds (WIP) standalone build script of OpenCV 2.4 Android SDK (with disabled OpenCV Android manager - it is outdated after OpenCV 3.x releases).
Soon I will add some hints/requirements/guidelines into description of this PR how to build updated Android SDK.

**Comment by apavlenko** [comment]
Below is an example how I build OpenCV 2.4 for Android SDK for arm-v7a on Windows:
```cmd
git clone --branch 2.4 --depth 1 https://github.com/opencv/opencv.git
mkdir opencv-build-android-arm-v7a
cd opencv-build-android-arm-v7a
set JAVA_HOME=C:\Program Files\Java\jdk1.8.0_102
set ANDROID_SDK=C:\Android\android-sdk-windows
set ANDROID_NDK=C:\Android\android-ndk-r10e
set MAKE_EXE=%ANDROID_NDK%\prebuilt\windows-x86_64\bin\make.exe
C:\Apps\cmake-3.5.2-win32-x86\bin\cmake.exe -G"MinGW Makefiles" -DCMAKE_MAKE_PROGRAM="%MAKE_EXE%" -DCMAKE_TOOLCHAIN_FILE=../opencv/platforms/android/android.toolchain.cmake -DANDROID_ABI="armeabi-v7a with NEON" -DANDROID_SDK_TARGET=21 -DANDROID_NATIVE_API_LEVEL=14 -DCMAKE_BUILD_WITH_INSTALL_RPATH=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_opencv_ocl=OFF -DBUILD_opencv_gpu=OFF -DBUILD_opencv_ts=OFF -DBUILD_ANDROID_EXAMPLES=OFF %* ../opencv
pause
%MAKE_EXE% install/strip
```
So it's possible to make similar builds for all the required arch-s and merge the build/install dir content.

**Comment by nisrulz** [comment]
@apavlenko Due by any idea know the equivalent for a MacOSX ?.
The problem I am facing is that I donot have cmake and the path `MAKE_EXE` , if I check the same for MacOSX I actually have something called a `gdbserver`.

I maybe missing something and would need a little help. I have always used the sdk as is, haven't ever tried building it from source before.

**Comment by alalek** [comment]
@nisrulz "cmake" is required to build OpenCV Android SDK.
There is [test build](http://pullrequest.opencv.org/buildbot/builders/precommit_pack_android/builds/149) for OpenCV 2.4 Android SDK. Could you download and check updated [package](http://pullrequest.opencv.org/buildbot/export/opencv_releases/precommit_pack_android/20160915-134521_149/) (artifact of mentioned build)? Does it work "as is" for you?

**Comment by nisrulz** [comment]
@alalek Thank you for clarifying for me, appreciated.

**The test build you referenced works for me.** 

## **For anyone else who is looking for a solution to a similar issue I am summarizing it below** 

OS : Mac OS X El Capitan
Open CV Android Package : [v2.4.13](http://pullrequest.opencv.org/buildbot/export/opencv_releases/precommit_pack_android/20160915-134521_149/)  [[Build Details](http://pullrequest.opencv.org/buildbot/builders/precommit_pack_android/builds/149)]

Compile your code into `.so`-> integrate it in your android app -> generate an apk.

Now to test the libpng version (just to verify it towards the end) , goto terminal and run the below command on your apk (for each arch, I am showing below for ARM abi)

```bash
unzip -p app-armeabi-debug.apk | strings | grep "libpng" 
```

you should get the below result

``` bash
png_get_libpng_ver
Application built with libpng-
libpng version 1.5.27 - May 26, 2016
libpng version 1.5.27 - May 26, 2016
```

Thats it , everything else works as is and libpng is updated to a version that google is not blocking for now.

Thanks to @alalek and @apavlenko for walking me through the process and helping me out in this.

**Comment by boazgarty** [comment]
@nisrulz Im trying to build arm64-v8a from your explanation.
the link v2.4.13 [Build Details] is not working. do you have solution for it?

**Comment by apavlenko** [comment]
@boazgarty, I checked my instruction above - it works for "arm64-v8a" on Windows and Linux (Ubuntu 16)

**Comment by boazgarty** [comment]
Thanks works I was missing the Linuk command

**Pull request merged**